runtime/compiler/env/j9method.cpp

```diff
@@ -9412,3 +9418,34 @@ TR_J9ByteCodeIlGenerator::packReferenceChainOffsets(TR::Node *node, std::vector<
    return;
    }
 #endif
+
+bool
+TR_ResolvedJ9Method::isFieldQType(int32_t cpIndex)
+   {
+   if (!TR::Compiler->om.areValueTypesEnabled() ||
+      (-1 == cpIndex))
+      return false;
+
+   J9VMThread *vmThread = fej9()->vmThread();
+   J9ROMFieldRef *ref = (J9ROMFieldRef *) (&romCPBase()[cpIndex]);
+   J9ROMNameAndSignature *nameAndSignature = J9ROMFIELDREF_NAMEANDSIGNATURE(ref);
+   J9UTF8 *signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSignature);
+   TR::VMAccessCriticalSection isFieldQType(fej9());
+
+   return vmThread->javaVM->internalVMFunctions->isNameOrSignatureQtype(signature);
+   }
+
+bool
+TR_ResolvedJ9Method::isFieldFlattened(TR::Compilation *comp, int32_t cpIndex, bool isStatic)
+   {
+   if (!TR::Compiler->om.areValueTypesEnabled() ||
+      (-1 == cpIndex))
+      return false;
+
+   J9VMThread *vmThread = fej9()->vmThread();
+   J9ROMFieldShape *fieldShape = NULL; 
+   TR_OpaqueClassBlock *containingClass = definingClassAndFieldShapeFromCPFieldRef(comp, cp(), cpIndex, isStatic, &fieldShape);
+   TR::VMAccessCriticalSection isFieldFlattened(fej9());
```

Comment:
I'm guessing the lock is needed here because we might need to decide to flatten the field for the first time and might need to update the VM in some way?

Reply:
This helper accesses `J9Class::flattenedClassCache`. It reads the cache. To prevent other threads from writing the cache, I think a lock would be required here.

Reply:
> To prevent other threads from writing the cache,

My comments above is not correct, clarified with @tajila, entires in `J9Class::flattenedClassCache` are only written during classload. They are effectively read only when being exposed to the JIT. I'll remove the lock next.